pkg/networkservice/common/discoverforwarder/server.go

```diff
@@ -93,6 +93,32 @@ func (d *discoverForwarderServer) Request(ctx context.Context, request *networks
 			return nil, errors.New("no candidates found")
 		}
 
+		segments := request.Connection.GetPath().GetPathSegments()
+		pathIndex := int(request.Connection.GetPath().Index)
+		datapathForwarder := ""
+		if len(segments) > pathIndex+1 {
+			datapathForwarder = segments[pathIndex+1].Name
+		}
+
+		for _, candidate := range nses {
+			if candidate.Name != datapathForwarder {
+				continue
+			}
+			u, err := url.Parse(candidate.Url)
+
+			if err != nil {
+				logger.Errorf("can not parse forwarder=%v url=%v error=%v", candidate.Name, u, err.Error())
+				return nil, errors.WithStack(err)
+			}
+
+			resp, err := next.Server(ctx).Request(clienturlctx.WithClientURL(ctx, u), request)
+			if err != nil {
+				continue
+			}
+			storeForwarderName(ctx, candidate.Name)
+			return resp, nil
+		}
```

Comment:
Looks like a code duplication from line 100.

```suggestion
for i, candidate := range nses {
    if candidate.Name == datapathForwarder {
         nses[0], nses[i] = nses[i], nses[0]
         break
    }
}
```